### PR TITLE
Enhance `ext/dom` stubs

### PR DESCRIPTION
**Pull request by Girgias**
Should be reviewed, and merged as individual commits.

@nielsdos could you have a quick look to see if I didn't make a wrong assumption somewhere?

### CODE REVIEW

**Comment by Girgias** [comment]
The 8.4 CI failure seems to be that the stubs don't actually understand what a covariant template is?

File: dom/dom_c.php [hunk 1]
```diff
@@ -1328,6 +1328,9 @@ public function registerNodeClass(
 /**
  * The DOMNodeList class
  * @link https://php.net/manual/en/class.domnodelist.php
+ *
+ * @template-covariant TNode as DOMNode
```
Comment (nielsdos) [inline dom/dom_c.php:1332]: Right, I just checked and it is unfortunately possible to get a DOMNameSpaceNode out of a DOMNodeList by using XPath. So this should be `TNode as DOMNode|DOMNameSpaceNode` if possible

File: dom/dom_c.php [hunk 2]
```diff
@@ -2478,7 +2481,7 @@ public function registerNamespace(
      * </p>
      * @param bool $registerNodeNS [optional] <p>The optional registerNodeNS can be specified to
      * disable automatic registration of the context node.</p>
-     * @return DOMNodeList|false a DOMNodeList containing all nodes matching
+     * @return DOMNodeList<DOMNode>|false a DOMNodeList containing all nodes matching
```
Comment (nielsdos) [inline dom/dom_c.php:2487]: Should be `DOMNodeList<DOMNode|DOMNameSpaceNode>|false`

File: dom/dom_c.php [hunk 3]
```diff
@@ -2505,7 +2508,7 @@ public function query(
      * <p>
      * The optional registerNodeNS can be specified to disable automatic registration of the context node.
      * </p>
-     * @return mixed a typed result if possible or a DOMNodeList
+     * @return mixed a typed result if possible or a DOMNodeList<DOMNode>
```
Comment (nielsdos) [inline dom/dom_c.php:2514]: Should be `DOMNodeList<DOMNode|DOMNameSpaceNode>|false`

**Review by nielsdos — COMMENTED** [review]
Just checked this. Almost right

File: dom/dom_n.php [hunk 4]
```diff
@@ -58,23 +65,31 @@ public function getIterator(): \Iterator {}
...
     public function getNamedItemNS(?string $namespace, string $localName): Entity|Notation|null {}
 
     public function count(): int {}
 
+    /** @return \Iterator<string, TDtdNode> */
     public function getIterator(): \Iterator {}
 }
 /**
  * @since 8.4
+ * @implements \IteratorAggregate<array-key, \DOM\Element>
```
Comment (veewee) [inline dom/dom_n.php:92]: implies that `getIterator()` returns `Element`s

```php
        /**
         * @return \Iterator<array-key, \DOM\Element>
         */
        public function getIterator(): \Iterator {}
```

File: dom/dom_n.php [hunk 5]
```diff
@@ -28,19 +28,26 @@ private function __construct() {}
...
+     * @return \Iterator<int, TNode>
+     */
     public function getIterator(): \Iterator {}
 
+    /** @return ?TNode */
     public function item(int $index): ?Node {}
 }
 /**
  * @since 8.4
+ * @implements \IteratorAggregate<array-key, \DOM\Attr>
```
Comment (veewee) [inline dom/dom_n.php:50]: implies that `getIterator()` returns `Attr`s

```php
        /**
         * @return \Iterator<array-key, \DOM\Attr>
         */
        public function getIterator(): \Iterator {}
```

**Comment by isfedorov** [comment]
Hey folks, thanks for the PR! Some of the tests are failing at the moment, due to not supported processing of `@template T as Type` notation. Right now tests just take template name and convert it later to `object` during comparison with signature.  So the tests should be updated to store now only template name, but also template types (`\StubTests\Model\PHPDocElement::collectTags`) and use an appropriate type later during comparison with function signature (`\StubTests\StubsTypeHintsTest::testClassesMethodsSignatureTypeHintsConformPhpDocInMethods`). any chance you could take a look and try to fix them? It’ll help get the PR merged faster. If not, no worries, we’ll sort it out on our side later, just with a bit of a delay.

**Comment by Girgias** [comment]
I will try to fix the tests, just wasn't sure what the approach to fix them should be. :)

**Comment by isfedorov** [comment]
@Girgias Thank you for fixing tests! All the changes looks good. Could you please take a look at inline comments and fix related items. After that PR can be merged.

**Comment by Girgias** [comment]
> @Girgias Thank you for fixing tests! All the changes looks good. Could you please take a look at inline comments and fix related items. After that PR can be merged.

Did you forget to submit the inline comments? I can't see any.

**Comment by Girgias** [comment]
@isfedorov, did you forget to submit the inline comments? I can't see any.

**Comment by isfedorov** [comment]
> @isfedorov, did you forget to submit the inline comments? I can't see any.

🤦 Indeed! Sorry, published review now.

File: dom/dom_c.php [hunk 6]
```diff
@@ -1328,6 +1328,9 @@ public function registerNodeClass(
 /**
  * The DOMNodeList class
  * @link https://php.net/manual/en/class.domnodelist.php
+ *
+ * @template-covariant TNode as DOMNode|DOMNameSpaceNode
```
[the remark this answers is not in the document]
Comment (Girgias) [reply dom/dom_c.php]: I was trying to handle it with: https://github.com/JetBrains/phpstorm-stubs/pull/1702/files#diff-5ff5df0ddd60476bd63dd67f5caac18378b3282d04a56ddace5a1b9d3f42d556R652-R676

**Comment by Girgias** [comment]
I don't really understand why there are now PGSQL failures on the 8.4 CI :(

**Comment by isfedorov** [comment]
@Girgias I believe the failures are not related to your changes but related to the updated PHP 8.4 docker image.
Screenshot 2024-12-12 at 22 29 14

```
1) StubTests\BaseConstantsTest::testConstants with data set "constant \LIBXML_NO_XXE" ('\LIBXML_NO_XXE')
Missing constant: const \LIBXML_NO_XXE = 8388608

2) StubTests\BaseConstantsTest::testConstants with data set "constant \PGSQL_TUPLES_CHUNK" ('\PGSQL_TUPLES_CHUNK')
Missing constant: const \PGSQL_TUPLES_CHUNK = 12

3) StubTests\BaseFunctionsTest::testFunctionsExist with data set "function \pg_set_chunked_rows_size" ('\pg_set_chunked_rows_size')
Missing function: function \pg_set_chunked_rows_size(Pgsql\Connection $connection, int $size){}
```
Tests are not bound to a strict version of PHP and use just latest minor updates of major versions so I suppose the related docker image was just updated at some moment in between your commits and new entities were added

**Comment by isfedorov** [comment]
@Girgias Could you please revert changes in files under .idea folder? Those are local specific changes and shouldn't be stored in VCS.

**Comment by Girgias** [comment]
> @Girgias Could you please revert changes in files under .idea folder? Those are local specific changes and shouldn't be stored in VCS.

Should the folder be added to `.gitignore`?

**Comment by isfedorov** [comment]
@Girgias It's already partially added. I wouldn't add the whole directory since some of its files may be good to share via VCS and also some of the files can contain both, content that does make sense to share and content that doesn't so every single case should be considered separately. Thank you for the fix!

**Pull request merged**